File: src/main/java/com/airwaves/airwavesweb/datastore/Cluster.java
package com.airwaves.airwavesweb.datastore;

import com.airwaves.airwavesweb.util.Util;
import com.google.cloud.firestore.DocumentReference;
import com.google.cloud.firestore.DocumentSnapshot;
import com.google.cloud.firestore.Firestore;
import com.google.cloud.firestore.SetOptions;

import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public class Cluster {

    private Firestore db;
    private DocumentReference clusterDocument;
    public static int maxUsers = 20;
    private int minUsers = 5;
    public static int writes = 0;
    public static int reads = 0;

    public static List<Cluster> getAll() {
        try {
            return Database.getDb().collection("cluster").get().get().getDocuments()
                    .stream().map(x -> new Cluster(x.getId())).collect(Collectors.toList());
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    public Cluster() {
        this.db = Database.getDb();
        this.clusterDocument = this.db.collection("cluster").document();
        this.clusterDocument.set(Map.of(
                "latitude", 0.0,
                "longitude", 0.0,
                "users", 0
        ));
    }

    public Cluster(String id) {
        this.db = Database.getDb();
        this.clusterDocument = this.db.collection("cluster").document(id);
    }

    public Cluster(List<User> users, double latitude, double longitude) {
        this.db = Database.getDb();
        this.clusterDocument = this.db.collection("cluster").document();
        this.clusterDocument.set(Map.of(
                "latitude", latitude,
                "longitude", longitude,
                "users", users.size()
        ));
    }

    public String getId() {
        return this.clusterDocument.getId();
    }

    public void set(Map<String, Object> data) {
        this.clusterDocument.set(data, SetOptions.merge());
        Cluster.writes++;
    }

    public void overwrite(Map<String, Object> data) {
        this.clusterDocument.set(data);
    }

    public DocumentSnapshot snapshot() {
        try {
            Cluster.reads++;
            return this.clusterDocument.get().get();
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    public double getLatitude() {
        var latitude = this.snapshot().getDouble("latitude");
        return latitude != null ? latitude : 0;
    }

    public double getLongitude() {
        var longitude = this.snapshot().getDouble("longitude");
        return longitude != null ? longitude : 0;
    }

    public List<User> getUsers() {
        try {
            return this.db.collection("user").whereEqualTo("cluster", this.getId()).get().get().getDocuments()
                    .stream().map(x -> new User(x.getId())).collect(Collectors.toList());
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    public void calculatePosition() {
        List<User> users = this.getUsers();
        var avgLatitude = users.stream().mapToDouble(User::getLatitude).average();
        var avgLongitude = users.stream().mapToDouble(User::getLongitude).average();

        if (avgLatitude.isPresent() && avgLongitude.isPresent()) {
            this.setLocation(avgLatitude.getAsDouble(), avgLongitude.getAsDouble());
        }
    }

    public void setLocation(double latitude, double longitude) {
        this.set(Map.of("latitude", latitude, "longitude", longitude));
    }

    public void splitOrMerge() {
        if (this.getUsers().size() > Cluster.maxUsers) {
            this.split();
        } else if (this.getUsers().size() < this.minUsers) {
            this.merge();
        }
    }

    private void split() {
        Cluster newCluster = new Cluster();
        List<User> users = this.getUsers();
        for (int i = 0; i < users.size() / 2; i++) {
            User user = users.get(i);
            user.setCluster(newCluster);
        }
    }

    private void merge() {
        Cluster closest = Util.findClosestCluster(this);
        if (closest != null) {
            for (User user : this.getUsers()) {
                user.setCluster(closest);
            }
        }
    }

    public void delete() {
        this.clusterDocument.delete();
    }
}

File: src/main/java/com/airwaves/airwavesweb/datastore/Cluster.kt
package com.airwaves.airwavesweb.datastore

import com.airwaves.airwavesweb.util.Util
import com.google.cloud.firestore.DocumentReference
import com.google.cloud.firestore.Firestore
import com.google.cloud.firestore.QueryDocumentSnapshot
import java.util.*
import java.util.stream.Collectors

class Cluster : Document() {
    private val db: Firestore? = null
    private val clusterDocument: DocumentReference? = null
    private val minUsers = 5
    val latitude: Double
        get() = getData()!!["latitude"] as Double

    val longitude: Double
        get() = getData()!!["longitude"] as Double

    val users: List<User>
        get() = try {
            db!!.collection("user").whereEqualTo("cluster", id).get().get().documents
                    .stream().map { x: QueryDocumentSnapshot -> User(x.id) }.collect(Collectors.toList())
        } catch (e: Exception) {
            throw RuntimeException(e)
        }

    fun calculatePosition() {
        val users = users
        val avgLatitude = users.stream().mapToDouble { obj: User -> obj.latitude }.average()
        val avgLongitude = users.stream().mapToDouble { obj: User -> obj.longitude }.average()
        if (avgLatitude.isPresent && avgLongitude.isPresent) {
            setLocation(avgLatitude.asDouble, avgLongitude.asDouble)
        }
    }

    fun setLocation(latitude: Double, longitude: Double) {
        getData()!!["latitude"] = latitude
        getData()!!["longitude"] = longitude
        save()
    }

    fun splitOrMerge() {
        if (users.size > maxUsers) {
            this.split()
        } else if (users.size < minUsers) {
            merge()
        }
    }

    private fun split() {
        val newCluster = Cluster()
        val users = users
        for (i in 0 until users.size / 2) {
            val user = users[i]
            user.cluster = newCluster
        }
    }

    private fun merge() {
        val closest = Util.findClosestCluster(this)
        if (closest != null) {
            for (user in users) {
                user.cluster = closest
            }
        }
    }

    fun delete() {
        clusterDocument!!.delete()
    }

    override fun collectionName(): String {
        return "cluster"
    }

    override fun defaultData(): HashMap<String, Any> {
        return HashMap<String, Any>(Map.of<String, Int>(
                "longitude", 0,
                "latitude", 0,
                "users", 0
        ))
    }

    companion object {
        var maxUsers = 20
        var writes = 0
        var reads = 0
        val all: List<Cluster>
            get() = getAll(Cluster::class.java, "cluster")
    }
}